src/preamble.js

```diff
@@ -164,27 +164,37 @@ function ccall (ident, returnType, argTypes, args, opts) {
     }
   }
   var ret = func.apply(null, cArgs);
-#if ASSERTIONS
 #if EMTERPRETIFY_ASYNC
+  if (opts && opts.callback) opts.async = true;
+#if ASSERTIONS
   if ((!opts || !opts.async) && typeof EmterpreterAsync === 'object') {
     assert(!EmterpreterAsync.state, 'cannot start async op with normal JS calling ccall');
   }
-  if (opts && opts.async) assert(!returnType, 'async ccalls cannot return values');
-#endif
+  if (opts && opts.async && returnType) assert(opts.callback, 'async ccalls need a callback to return values');
+#endif
+  if (opts && opts.async && EmterpreterAsync.state) {
+    EmterpreterAsync.restartFunc = func;
+    EmterpreterAsync.asyncFinalizers.push(function(ret) {
+      EmterpreterAsync.restartFunc = null;
+      if (stack !== 0) stackRestore(stack);
+      if (opts.callback) {
+        if (returnType === 'string') ret = Pointer_stringify(ret);
+        else if (returnType === 'boolean') ret = Boolean(ret);
+        opts.callback(ret);
+      }
+    });
+    return;
+  }
 #endif
   if (returnType === 'string') ret = Pointer_stringify(ret);
   else if (returnType === 'boolean') ret = Boolean(ret);
-  if (stack !== 0) {
+  if (stack !== 0) stackRestore(stack);
 #if EMTERPRETIFY_ASYNC
-    if (opts && opts.async) {
-      EmterpreterAsync.asyncFinalizers.push(function() {
-        stackRestore(stack);
-      });
-      return;
-    }
-#endif
-    stackRestore(stack);
+  if (opts && opts.callback) {
+    opts.callback(ret);
```

Comment:
This is for if the function finishes without needing to suspend for an async operation, right? I think this should delay the calling of the callback for consistency. `setTimeout` has a minimum delay of 4ms but I don't think it would be worth the code cost to implement anything with less delay.

Reply:
4ms is a lot, if you do just 250 calls that's already a whole second. If we really have to delay the callback, we can just as well use promises, where this delay is already built in (and not 4ms).
But in my opinion, if someone who uses ccall needs this delayed behavior, they can just implement it themselves.

Reply:
Right, I guess the delay could easily become a problem.

This might actually not be an excessive code weight addition:

    Promise.resolve(ret).then(opts.callback)

Or ccall itself could return a promise.